### ISSUE TITLE
Confusion of discriminator loss 

### ISSUE THREAD

**Issue by Opdoop** [PanZiqiAI/disentangled-information-bottleneck#1]
In `Algorithm 1`, <img src="https://render.githubusercontent.com/render/math?math=\mathcal{L}_{\phi}"> is calculated as (equation 83):

<img src="https://render.githubusercontent.com/render/math?math=\mathcal{L}_{\phi}=\frac{1}{|\mathcal{B}|} \sum_{i \in \mathcal{B}}-\log \left(1-W\left(s_{i}, t_{i}\right)\right)-\log W\left(t_{\pi(i)}, s_{\pi(i)}\right)">

But in the implementation, the loss is calculated using these lines.
https://github.com/PanZiqiAI/disentangled-information-bottleneck/blob/5c1790857b90630880082b945be582638f234568/shared_libs/utils/criterions.py#L86-L87

Why does this inconsistent happen? Correct me if I am wrong.

**Comment by PanZiqiAI** [comment]
Hello, <img src="https://render.githubusercontent.com/render/math?math=\mathcal{L}_{\phi}"> (`Eq.(83)` in `Algorithm 1`) is the _Binary Cross Entropy_ loss of the vanilla GAN, while in the implementation a modified _hinge_ loss is used. Any adversarial loss, such as the vanilla _BCE_ loss, _hinge_ loss, _WGAN_ loss, etc., can be used for the adversarial training in `Algorithm 1`.

**Comment by PanZiqiAI** [comment]
<img src="https://render.githubusercontent.com/render/math?math=\mathcal{L}_{\phi}"> (`Eq.(83)` in `Algorithm 1`) is the _Binary Cross Entropy_ loss of the vanilla GAN, while in the implementation a modified _hinge_ loss is used. Any adversarial loss, such as the vanilla _BCE_ loss, _hinge_ loss, _WGAN_ loss, etc., can be used for the adversarial training in `Algorithm 1`.

**Comment by Opdoop** [comment]
Thank you for your prompt reply! 
I'm not familiar with GAN. This modified hinge loss looks like a trick. The following question is why you use this modified hinge loss in the implementation? Have you tried other losses? Does the performance of the discriminator(or say `I (S; T)`) sensitive to those losses?

**Comment by PanZiqiAI** [comment]
I found this loss work and I didn't try other losses. The adversarial training manner involving a discriminator is very sensitive to not only losses, but also almost every experimental setups. To train a discriminator successfully, many tricks are usually used,  and how to stabilize training still remains a major research problem of training GANs (adversarial training).

**Comment by Opdoop** [comment]
Thanks again! Your explanation is really helpful!
Another related question is how to judge the discriminator is good enough?
I run the `python train_disenib.py --dataset=mnist --desc=quant/mnist/disenib` command.

![image](https://user-images.githubusercontent.com/21202514/144426600-a39a686f-1cdc-4a28-b203-2c9faede91d7.png)

The `est_real` and `est_fake` loss(above picture) looks quite weird. How did you decide when to stop the training?

**Comment by PanZiqiAI** [comment]
Taking _Generative Adversarial Nets_ (GANs) as an example, which consists of a _generator_ <img src="https://render.githubusercontent.com/render/math?math=G"> and a _discriminator_ <img src="https://render.githubusercontent.com/render/math?math=D">, where the generator <img src="https://render.githubusercontent.com/render/math?math=G"> aims to generate high quality images that the discriminator <img src="https://render.githubusercontent.com/render/math?math=D"> cannot distinguish from real images (comes from a training dataset), and the discriminator <img src="https://render.githubusercontent.com/render/math?math=D"> aims to distinguish fake images (generated by the generator <img src="https://render.githubusercontent.com/render/math?math=G">) and real images. Assume that <img src="https://render.githubusercontent.com/render/math?math=D"> assigns <img src="https://render.githubusercontent.com/render/math?math=1"> for real images and assigns <img src="https://render.githubusercontent.com/render/math?math=0"> for fake images, then <img src="https://render.githubusercontent.com/render/math?math=G"> are trained to generate images <img src="https://render.githubusercontent.com/render/math?math=x"> such that <img src="https://render.githubusercontent.com/render/math?math=D\left(x\right)"> is as close to <img src="https://render.githubusercontent.com/render/math?math=1"> as possible, and <img src="https://render.githubusercontent.com/render/math?math=D"> are trained such that <img src="https://render.githubusercontent.com/render/math?math=D\left(x\right)=0"> and <img src="https://render.githubusercontent.com/render/math?math=D\left(y\right)=1"> (<img src="https://render.githubusercontent.com/render/math?math=y"> is a real image from the dataset). So when given a fake image <img src="https://render.githubusercontent.com/render/math?math=x"> generated by the generator <img src="https://render.githubusercontent.com/render/math?math=G">, <img src="https://render.githubusercontent.com/render/math?math=D\left(x\right)\approx 0.5">, namely the discriminator cannot confidently distinguish whether it is a real or fake image, the whole adversarial system achieves the Nash Equilibrium and we can say it is good. So in our case, we expect both the `est_real` and `est_fake` losses are close to <img src="https://render.githubusercontent.com/render/math?math=0">, which means that fake samples are difficult to be distinguished from real samples.

**Comment by Opdoop** [comment]
If I understand correctly, we expect both the `est_real` and `est_fake` losses are close to 0 in which the discriminator cannot confidently distinguish `q(s; t)` and `q(s)q(t)`. At that time, we minimized `I(S;T)` term.
Use the above loss curve as a concrete example. 
At step 2M, there is a sharp change of `est_real` and `est_fake`. Both of the two losses close to 0 very quickly. That's the time we use discriminator loss to update encoders. Specific to the code, these lines below start to work.
https://github.com/PanZiqiAI/disentangled-information-bottleneck/blob/5c1790857b90630880082b945be582638f234568/methods/disenib_fc/disenib.py#L103-L111
My questions are: 
1. What causes these losses to deviate from 0 after 5M steps? 
2. Why don't we stop training at around 4M to 5M steps but keep training until it extends 12M steps? Don't we break Nash Equilibrium because losses are not close to 0 anymore?

**Comment by PanZiqiAI** [comment]
This is hard to explain, and I don't know why it starts to deviate from 0 neither. The Nash Equilibrium is an ideal result in theory, and in my practice of training GANs, I found the generator can also produce high quality images when the discriminator are not confused, namely confidently distinguish real and fake images. I didn't tune the `training steps` parameter carefully.

**Comment by Opdoop** [comment]
A big thank you for your patient answer! 
I have one last question. The references at the implementation part of the DisenIB principle are a little vague. Is using adversarial training for minimizing mutual information is your original? Or it's derived from other mutual information estimate methods? I wish to dig deeper into mutual information estimation along this way. Any information is welcome~
Thanks again!😀

**Comment by PanZiqiAI** [comment]
The `FactorVAE` [1] also uses adversarial training to minimizing <img src="https://render.githubusercontent.com/render/math?math=D_{\mathrm{KL}}\left[p\left(z\right)\parallel \prod_{j} p\left(z_j\right)\right]">, which is also using adversarial training for minimizing mutual information since KL diverigence is essential a mutual information. 

[1] Kim H , Mnih A . Disentangling by Factorising. ICML 2018.
